### Part-5: Data Analysis with SQL

##### Overview of the DataSet

SpaceX has gained worldwide attention for a series of historic milestones. 

It is the only private company ever to return a spacecraft from low-earth orbit, which it first accomplished in December 2010.
SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars wheras other providers cost upward of 165 million dollars each, much of the savings is because Space X can reuse the first stage. 


Therefore if we can determine if the first stage will land, we can determine the cost of a launch. 

This information can be used if an alternate company wants to bid against SpaceX for a rocket launch.

This dataset includes a record for each payload carried during a SpaceX mission into outer space.


In [35]:
!pip install sqlalchemy


### Connect to the database

Let us first load the SQL extension and establish a connection with the database


In [36]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [37]:
import csv, sqlite3
import pandas as pd

con = sqlite3.connect("my_data1.db")
cur = con.cursor() # to work with queries , but i will use python function to query 

In [38]:
%sql sqlite:///my_data1.db

Traceback (most recent call last):
  File "/Users/rabinkarki/opt/anaconda3/lib/python3.9/site-packages/sql/magic.py", line 203, in execute
    conn.internal_connection.rollback()
AttributeError: 'Connection' object has no attribute 'rollback'

Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys(['sqlite:///my_data1.db'])


In [39]:

df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_2/data/Spacex.csv")
df.to_sql("SPACEXTBL", con, if_exists='replace', index=False,method="multi")
 

/Users/rabinkarki/opt/anaconda3/lib/python3.9/site-packages/pandas/core/generic.py:2605: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  sql.to_sql(


**Note:This below code is added to remove blank rows from table**


In [40]:
%sql create table SPACEXTABLE as select * from SPACEXTBL where Date is not null

 * sqlite:///my_data1.db
Traceback (most recent call last):
  File "/Users/rabinkarki/opt/anaconda3/lib/python3.9/site-packages/sql/magic.py", line 203, in execute
    conn.internal_connection.rollback()
AttributeError: 'Connection' object has no attribute 'rollback'

Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys(['sqlite:///my_data1.db'])




Now write and execute some SQL queries to find some meaningful insights . 
=

### Task 1
##### Display the names of the unique launch sites  in the space mission


In [41]:
q = pd.read_sql('select distinct Launch_Site from SPACEXTBL', con)
q

Launch_Site
0   CCAFS LC-40
1   VAFB SLC-4E
2    KSC LC-39A
3  CCAFS SLC-40

As we can see that  there are 4 unique launch sites that Falcon 9 make their launches.We used distinct keyword to select the filter data.


### Task 2


#####  Display 5 records where launch sites begin with the string 'CCA' 


In [42]:
lunch_record = pd.read_sql("select * from SPACEXTBL where Launch_Site like 'CCA%' limit 5", con)
lunch_record

Date Time (UTC) Booster_Version  Launch_Site  \
0    6/4/2010   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1   12/8/2010   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2  22/05/2012    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3   10/8/2012    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4    3/1/2013   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  PAYLOAD_MASS__KG_  \
0               Dragon Spacecraft Qualification Unit                  0   
1  Dragon demo flight C1, two CubeSats, barrel of...                  0   
2                              Dragon demo flight C2                525   
3                                       SpaceX CRS-1                500   
4                                       SpaceX CRS-2                677   

       Orbit         Customer Mission_Outcome      Landing_Outcome  
0        LEO           SpaceX         Success  Failure (parachute)  
1  LEO (ISS)  NASA (COTS) NRO         Success  Failure (parachute)  
2  LEO (ISS)      NASA (COTS)         Success           No attempt  
3  LEO (ISS)       NASA (CRS)         Success           No attempt  
4  LEO (ISS)       NASA (CRS)         Success           No attempt

We used "like" operator to get the launch sites informations whose name starts with "CCA" .

### Task 3




##### Display the total payload mass carried by boosters launched by NASA (CRS)


In [43]:
total_payload = pd.read_sql("select sum(PAYLOAD_MASS__KG_) from SPACEXTBL where Customer='NASA (CRS)' ", con)
total_payload

sum(PAYLOAD_MASS__KG_)
0                   45596

As we can see that total mass caried by Boosters launched by NASA is 45596 kg. We used sql "sum" function . 

### In addition ,
Let's List the names of customers who have had the most launches.


In [50]:
query ="""SELECT Customer, COUNT(*) AS LaunchCount
FROM SPACEXTBL
GROUP BY Customer
ORDER BY LaunchCount DESC
LIMIT 1"""
result = pd.read_sql(query, con)
result


Customer  LaunchCount
0  NASA (CRS)           20

In [54]:
query ="""SELECT Customer, COUNT(*) AS LaunchCount
FROM SPACEXTBL
GROUP BY Customer
ORDER BY LaunchCount DESC limit 10"""
result = pd.read_sql(query, con)
result


Customer  LaunchCount
0              NASA (CRS)           20
1                  SpaceX           13
2  Iridium Communications            7
3                     SES            5
4                 Thaicom            2
5                 Telesat            2
6     SpaceX, Planet Labs            2
7  SKY Perfect JSAT Group            2
8                 Orbcomm            2
9                 AsiaSat            2

Its indeed NASA

### Task 4




##### Display average payload mass carried by booster version F9 v1.1


In [44]:

avg_payloadd_mass = pd.read_sql("select avg(PAYLOAD_MASS__KG_) from SPACEXTBL where Booster_Version = 'F9 v1.1'", con)
avg_payloadd_mass


avg(PAYLOAD_MASS__KG_)
0                  2928.4

We used the similar aggrete function as above but "avg" this time.

### Task 5

##### List the date when the first succesful landing outcome in ground pad was acheived.


_Hint:Use min function_ 


In [45]:
fsl_date = pd.read_sql("select min(Date) as Minimum_Date from SPACEXTBL where Landing_Outcome ='Success (ground pad)'", con)
fsl_date

Minimum_Date
0     1/8/2018

we used "min" function this time to get the first date that appeared after where clause . 

### Task 6
 This is important task because as we saw in the previus section in EDA Notebook, most of the booster carried payload mass betweeen 4000 kg and 6000 kg." 
 Let's Find out .
 
##### List the names of the boosters which have success in drone ship and have payload mass greater than 4000 but less than 6000


In [46]:
booster_name= pd.read_sql("select distinct  Booster_Version from SPACEXTBL where Landing_Outcome = 'Success (drone ship)' and PAYLOAD_MASS__KG_ between 4000 and 6000 ", con)
booster_name

Booster_Version
0     F9 FT B1022
1     F9 FT B1026
2  F9 FT  B1021.2
3  F9 FT  B1031.2

As we can see that there are 4 distinct falcon 9 booster version who carried payload mass between 4000 kg and 6000 kg.
Here ,
we select distinct name of the booster versions where landing outcome is success in drone ship and payload mass between 4k and 6k kg.
"and", "between" operaotrs helps us to interset the data and find the between values .

#### Let's also find out different payload types 

In [57]:
query  = """SELECT Payload, COUNT(*) AS PayloadCount
FROM SPACEXTBL
GROUP BY Payload
ORDER BY PayloadCount DESC
LIMIT 5
"""
result = pd.read_sql(query, con)
result

Payload  PayloadCount
0                                          Zuma             1
1                        Turkmen 52 / MonacoSAT             1
2  Transiting Exoplanet Survey Satellite (TESS)             1
3                                     Thaicom 8             1
4                                     Thaicom 6             1

### Task 7




##### List the total number of successful and failure mission outcomes


In [49]:

# both success and failure mission outcome in the same table using substring method
# and group by the first selected clause from the Misson_outcome table
sf = pd.read_sql("select substr(Mission_Outcome , 1, 7) as Mission_Outcomes, count(*) as count from SPACEXTBL group by 1 ", con)
sf


Mission_Outcomes  count
0          Failure      1
1          Success    100

### Let's also calculate the success rate for each launch sites

In [63]:
result = pd.read_sql("""SELECT Launch_Site, COUNT(*) AS TotalLaunches, SUM(CASE WHEN Mission_Outcome = 'Success' THEN 1 ELSE 0 END) AS SuccessfulLaunches
FROM SPACEXTBL
WHERE Mission_Outcome IN ('Success', 'Failure')
GROUP BY Launch_Site""", con)
result


Launch_Site  TotalLaunches  SuccessfulLaunches
0   CCAFS LC-40             25                  25
1  CCAFS SLC-40             32                  32
2    KSC LC-39A             25                  25
3   VAFB SLC-4E             16                  16

### Task 8



##### List the   names of the booster_versions which have carried the maximum payload mass. Use a subquery


In [ ]:
bv = pd.read_sql("select distinct Booster_Version from SPACEXTBL where PAYLOAD_MASS__KG_ = (select max(PAYLOAD_MASS__KG_) from SPACEXTBL)", con)
bv

Booster_Version
0    F9 B5 B1048.4
1    F9 B5 B1049.4
2    F9 B5 B1051.3
3    F9 B5 B1056.4
4    F9 B5 B1048.5
5    F9 B5 B1051.4
6    F9 B5 B1049.5
7   F9 B5 B1060.2 
8   F9 B5 B1058.3 
9    F9 B5 B1051.6
10   F9 B5 B1060.3
11  F9 B5 B1049.7

From the above query, we can see that the maximum payload mass carried by different booster version of falcon9 .

### Task 9


##### List the records which will display the month names, failure landing_outcomes in drone ship ,booster versions, launch_site for the months in year 2015.

**Note: SQLLite does not support monthnames. So you need to use  substr(Date, 6,2) as month to get the months and substr(Date,0,5)='2015' for year.**


In [ ]:
a = pd.read_sql("""
    SELECT
        CASE
            WHEN substr(Date, 6, 2) = '01' THEN 'January'
            WHEN substr(Date, 6, 2) = '02' THEN 'February'
            WHEN substr(Date, 6, 2) = '03' THEN 'March'
            WHEN substr(Date, 6, 2) = '04' THEN 'April'
            WHEN substr(Date, 6, 2) = '05' THEN 'May'
            WHEN substr(Date, 6, 2) = '06' THEN 'June'
            WHEN substr(Date, 6, 2) = '07' THEN 'July'
            WHEN substr(Date, 6, 2) = '08' THEN 'August'
            WHEN substr(Date, 6, 2) = '09' THEN 'September'
            WHEN substr(Date, 6, 2) = '10' THEN 'October'
            WHEN substr(Date, 6, 2) = '11' THEN 'November'
            WHEN substr(Date, 6, 2) = '12' THEN 'December'
            ELSE 'Unknown'
        END AS MonthName,
        Landing_Outcome AS FailureLandingOutcomes,
        Booster_Version,
        Launch_Site
    FROM SPACEXTBL
    WHERE substr(Date, 1, 4) = '2015'
    AND Landing_Outcome LIKE '%drone ship%';
""", con)

a


MonthName  FailureLandingOutcomes Booster_Version  Launch_Site
0   October    Failure (drone ship)   F9 v1.1 B1012  CCAFS LC-40
1     April    Failure (drone ship)   F9 v1.1 B1015  CCAFS LC-40
2      June  Precluded (drone ship)   F9 v1.1 B1018  CCAFS LC-40

Above we used switch statements to select different months based on substring values . As we can see that three months were not that great time for Falcon9 in 2015. 

### Task 10




##### Rank the count of landing outcomes (such as Failure (drone ship) or Success (ground pad)) between the date 2010-06-04 and 2017-03-20, in descending order.


In [ ]:
a = pd.read_sql("""
    SELECT
        Landing_Outcome,
        COUNT(*) AS Count
    FROM SPACEXTBL
    WHERE Date BETWEEN '2010-06-04' AND '2017-03-20'
    GROUP BY Landing_Outcome
    ORDER BY Count DESC
""", con)

a


Landing_Outcome  Count
0              No attempt     10
1    Success (ground pad)      5
2    Success (drone ship)      5
3    Failure (drone ship)      5
4      Controlled (ocean)      3
5    Uncontrolled (ocean)      2
6  Precluded (drone ship)      1
7     Failure (parachute)      1

Note: The order of execution of the above query as follow:  FROM -> GROUP BY -> SELECT -> WHERE -> ORDER BY.

### Conclusion: 


From the above SQL Queries we retreived some important insights from spacex table about falcon 9 launch data. Understanding which booster versions are used to carry heavy payload and success rate between certain date helps us to understand more about Falcon 9 launches.

### Author: 
Rabin Karki 
